packages/nextjs/components/scaffold-eth/Contract/ContractUI.tsx

```diff
@@ -1,82 +1,194 @@
-  deployedContractData: { address: string; abi: Abi };
+  initialContractData: { address: string; abi: Abi };
 };
 
 const mainNetworks = getTargetNetworks();
 
 /**
  * UI component to interface with deployed contracts.
  **/
-export const ContractUI = ({ className = "", deployedContractData }: ContractUIProps) => {
+export const ContractUI = ({ className = "", initialContractData }: ContractUIProps) => {
   const [refreshDisplayVariables, triggerRefreshDisplayVariables] = useReducer(value => !value, false);
   const mainChainId = useAbiNinjaState(state => state.mainChainId);
   const mainNetwork = mainNetworks.find(network => network.id === mainChainId);
   const networkColor = useNetworkColor(mainNetwork);
 
+  const updateUrlWithSelectedMethods = (selectedMethods: string[]) => {
+    const currentQuery = new URLSearchParams(window.location.search);
+    if (selectedMethods.length > 0) {
+      currentQuery.set("methods", selectedMethods.join(","));
+    } else {
+      currentQuery.delete("methods");
+    }
+    const newPath = `/${initialContractData.address}/${mainChainId}`;
+
+    router.push({ pathname: newPath, query: currentQuery.toString() }, undefined, { shallow: true });
+  };
+
+  // include all functions with inputs
+  const methodsWithInputs = initialContractData.abi.filter(
+    (method): method is AbiFunction =>
+      method.type === "function" &&
+      "inputs" in method &&
+      (method.inputs.length > 0 || method.stateMutability === "payable"),
+  );
+
+  // include non-payable functions with no inputs
+  const [abi, setAbi] = useState<AbiFunction[]>(
+    initialContractData.abi.filter(
+      (method): method is AbiFunction =>
+        method.type === "function" &&
+        "inputs" in method &&
+        method.inputs.length === 0 &&
+        method.stateMutability !== "payable",
+    ),
+  );
+
+  const handleMethodSelect = (methodName: string) => {
+    const methodToAdd = initialContractData.abi.find(
+      method => method.type === "function" && "name" in method && method.name === methodName,
+    ) as AbiFunction | undefined; // Cast it to AbiFunction | undefined
+
+    if (methodToAdd && !abi.some(method => method.name === methodName)) {
+      const updatedAbi = [...abi, methodToAdd];
+      setAbi(updatedAbi);
+      updateUrlWithSelectedMethods(updatedAbi.map(m => m.name));
+    }
+  };
+
+  const removeMethod = (methodName: string) => {
+    const updatedAbi = abi.filter(fn => fn.name !== methodName);
+    setAbi(updatedAbi);
+    updateUrlWithSelectedMethods(updatedAbi.map(m => m.name));
+  };
+
+  useEffect(() => {
+    const selectedMethodNames = (router.query.methods as string)?.split(",") || [];
+    const selectedMethods = initialContractData.abi.filter(
+      method => method.type === "function" && "name" in method && selectedMethodNames.includes(method.name),
+    ) as AbiFunction[]; // Cast it to AbiFunction[]
+    setAbi(selectedMethods);
+  }, [router.query.methods, initialContractData.abi]);
```

Comment:
`router.query.methods` lets remove it from dependency, I think eslint also gives us a nice warning. 

If someone changes the url manually he will need to hit "enter" (which will reaload the page) and this useEffect will run and set state properly